File: src/main/java/io/github/takenoha/towerdefense/tactical/TacticalEffectCache.java
package io.github.takenoha.towerdefense.tactical;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.UUID;

/** In-memory cache of compiled effects; it never reads persistence on a combat hot path. */
public final class TacticalEffectCache implements TacticalEffectSnapshotProvider {
    private final TacticalBuildStateProvider stateProvider;
    private final TacticalEffectCompiler compiler;
    private final Map<UUID, TacticalEffectSnapshot> snapshots = new HashMap<>();

    public TacticalEffectCache(TacticalBuildStateProvider stateProvider) {
        this(stateProvider, new TacticalEffectCompiler());
    }

    public TacticalEffectCache(
            TacticalBuildStateProvider stateProvider,
            TacticalEffectCompiler compiler) {
        this.stateProvider = Objects.requireNonNull(stateProvider, "stateProvider");
        this.compiler = Objects.requireNonNull(compiler, "compiler");
    }

    /** Rebuilds one cache entry from the selected, versioned definition snapshot. */
    public synchronized void rebuild(UUID defenseId) {
        Objects.requireNonNull(defenseId, "defenseId");
        // Remove first so a provider/compiler failure cannot leave stale effects active.
        snapshots.remove(defenseId);
        Optional<TacticalBuildSelectionView> selected = stateProvider.findActiveByDefense(defenseId);
        if (selected == null || selected.isEmpty()) {
            return;
        }
        snapshots.put(defenseId, compiler.compile(selected.orElseThrow()));
    }

    /** Installs a supplied snapshot for recovery/tests without adding a persistence dependency. */
    public synchronized void rebuild(UUID defenseId, TacticalBuildSelectionView selected) {
        Objects.requireNonNull(defenseId, "defenseId");
        snapshots.remove(defenseId);
        snapshots.put(defenseId, compiler.compile(Objects.requireNonNull(selected, "selected")));
    }

    public synchronized void invalidate(UUID defenseId) {
        Objects.requireNonNull(defenseId, "defenseId");
        snapshots.remove(defenseId);
    }

    public synchronized void clear() {
        snapshots.clear();
    }

    @Override
    public synchronized TacticalEffectSnapshot currentForDefense(UUID defenseId) {
        Objects.requireNonNull(defenseId, "defenseId");
        return snapshots.getOrDefault(defenseId, EmptyTacticalEffectSnapshot.INSTANCE);
    }

    public synchronized int size() {
        return snapshots.size();
    }
}

File: src/main/kotlin/io/github/takenoha/towerdefense/tactical/TacticalEffectCache.kt
package io.github.takenoha.towerdefense.tactical

import java.util.HashMap
import java.util.Objects
import java.util.Optional
import java.util.UUID

/** In-memory cache of compiled effects; it never reads persistence on a combat hot path. */
class TacticalEffectCache(
    private val stateProvider: TacticalBuildStateProvider,
    private val compiler: TacticalEffectCompiler,
) : TacticalEffectSnapshotProvider {
    constructor(stateProvider: TacticalBuildStateProvider) : this(
        stateProvider,
        TacticalEffectCompiler(),
    )

    private val snapshots: MutableMap<UUID, TacticalEffectSnapshot> = HashMap()

    /** Rebuilds one cache entry from the selected, versioned definition snapshot. */
    @Synchronized
    fun rebuild(defenseId: UUID) {
        Objects.requireNonNull(defenseId, "defenseId")
        // Remove first so a provider/compiler failure cannot leave stale effects active.
        snapshots.remove(defenseId)
        val selected: Optional<TacticalBuildSelectionView>? =
            stateProvider.findActiveByDefense(defenseId)
        if (selected == null || selected.isEmpty()) {
            return
        }
        snapshots[defenseId] = compiler.compile(selected.orElseThrow())
    }

    /** Installs a supplied snapshot for recovery/tests without adding a persistence dependency. */
    @Synchronized
    fun rebuild(defenseId: UUID, selected: TacticalBuildSelectionView) {
        Objects.requireNonNull(defenseId, "defenseId")
        snapshots.remove(defenseId)
        snapshots[defenseId] = compiler.compile(
            Objects.requireNonNull(selected, "selected"),
        )
    }

    @Synchronized
    fun invalidate(defenseId: UUID) {
        Objects.requireNonNull(defenseId, "defenseId")
        snapshots.remove(defenseId)
    }

    /** Returns whether a selected tactical build is currently cached for the defense. */
    @Synchronized
    fun hasSelectedBuild(defenseId: UUID): Boolean {
        Objects.requireNonNull(defenseId, "defenseId")
        return snapshots.containsKey(defenseId)
    }

    @Synchronized
    fun clear() {
        snapshots.clear()
    }

    @Synchronized
    override fun currentForDefense(defenseId: UUID): TacticalEffectSnapshot {
        Objects.requireNonNull(defenseId, "defenseId")
        return snapshots.getOrDefault(defenseId, EmptyTacticalEffectSnapshot.INSTANCE)
    }

    @Synchronized
    fun size(): Int = snapshots.size
}
